### PR TITLE
[TRTLLM-11189][fix] Fix TeaCache broken caching for FLUX.1 and FLUX.2

### PR DESCRIPTION
**Pull request by karljang**
## Release Notes

* **Refactor**
  * Improved timestep and guidance embedding computation in FLUX.1 visual generation pipeline.
  * Enhanced TeaCache configuration to automatically select model-appropriate coefficients based on checkpoint identification.
  * Updated TeaCache coefficients for FLUX.2.

* **Bug Fixes**
  * Fixed TeaCache distance accumulation behavior to improve caching efficiency.

## Description

Fix bugs that made TeaCache produce garbage images (FLUX.2) or use wrong distance signals (FLUX.1)
- **`teacache.py` (common)**: Remove spurious 0.95 decay on cache hits — not in original paper or feature branch, creates positive feedback loop increasing hit rate by ~30-50%
- **`pipeline_flux.py` (FLUX.1)**: Fix timestep embedder calling `time_text_embed(timestep)` which always raised TypeError (needs 3 args: timestep, guidance, pooled_projection), silently falling back to raw timestep values; set polynomial coefficients directly bypassing broken path matching in `_setup_teacache()`
- **`pipeline_flux2.py` (FLUX.2)**: Fix wrong polynomial coefficients (was using FLUX.1's values); set `use_ret_steps=False` (was defaulting to `True`); set polynomial coefficients directly bypassing broken path matching in `_setup_teacache()`
- **Root cause of coefficient loading failure (common)**: `_setup_teacache()` in `pipeline.py` matches `pretrained_config._name_or_path` against coefficient keys, but `pretrained_config` is a `SimpleNamespace` with no `_name_or_path` attribute — polynomial always defaulted to identity `[1.0, 0.0]`

| Model | Before PR | After PR | Impact |
|-------|-----------|----------|--------|
| **FLUX.2-dev** (50 steps, thresh=0.2) | 94% hit — garbage images | 72% hit — good quality | Major fix |
| **FLUX.1-dev** (50 steps, thresh=0.3) | 48% hit — wrong distance signal | 60% hit — correct embedder + polynomial | Embedder + coefficients fixed |
| **WAN 2.1 T2V 1.3B** (50 steps, thresh=0.2) | 74% hit | 72% hit | Minimal (decay removal only) |

## Test Coverage

- [x] FLUX.2-dev on B200: hit rate 94% → 72%, garbage images → good quality
- [x] FLUX.1-dev on B200: hit rate 48% (wrong signal) → 60% (correct embedder + polynomial)
- [x] WAN 2.1 T2V 1.3B on B200: hit rate 74% → 72%, minimal change (only decay removal applies)
- [x] Visual inspection of output images vs baseline (no TeaCache)

## PR Checklist

Please review the following before submitting your PR:
- PR description clearly explains what and why. If using CodeRabbit's summary, please make sure it makes sense.
- PR Follows [TRT-LLM CODING GUIDELINES](https://github.com/NVIDIA/TensorRT-LLM/blob/main/CODING_GUIDELINES.md) to the best of your knowledge.
- Test cases are provided for new code paths (see [test instructions](https://github.com/NVIDIA/TensorRT-LLM/tree/main/tests#1-how-does-the-ci-work))
- Any new dependencies have been scanned for license and vulnerabilities
- [CODEOWNERS](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/CODEOWNERS) updated if ownership changes
- Documentation updated as needed
- Update [tava architecture diagram](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/tava_architecture_diagram.md) if there is a significant design change in PR.
- The reviewers assigned automatically/manually are appropriate for the PR.

- [x] Please check this after reviewing the above items as appropriate for this PR.

## GitHub Bot Help

To see a list of available CI bot commands, please comment `/bot help`.

### CODE REVIEW

**Comment by karljang** [comment]
@chang-l , @o-stoner, could you please review this PR~, thank you!

File: tensorrt_llm/_torch/visual_gen/models/flux/pipeline_flux2.py [hunk 1]
```diff
@@ -301,8 +301,14 @@ def post_load_weights(self) -> None:
                 )
             )
 
-            # Enable TeaCache with FLUX.2-specific coefficients
-            self._setup_teacache(self.transformer, coefficients=FLUX2_TEACACHE_COEFFICIENTS)
+            # FLUX.2 uses standard warmup mode (not ret_steps)
+            self.model_config.teacache.use_ret_steps = False
```
Comment (chang-l) [inline tensorrt_llm/_torch/visual_gen/models/flux/pipeline_flux2.py]: Why we override user config `use_ret_steps` here?

File: tensorrt_llm/_torch/visual_gen/models/flux/pipeline_flux2.py [hunk 2]
```diff
@@ -44,11 +44,11 @@
 
 from .transformer_flux2 import Flux2Transformer2DModel
 
-# TeaCache coefficients for FLUX.2
+# TeaCache coefficients for FLUX.2 (different from FLUX.1)
 FLUX2_TEACACHE_COEFFICIENTS = {
     "dev": {
```
[the remark this answers is not in the document]
Comment (karljang) [reply tensorrt_llm/_torch/visual_gen/models/flux/pipeline_flux2.py]: We don't have coefficients for Klein, I'll just add a debug log message:  "No TeaCache coefficients for variant %s, using dev defaults"

**Review by chang-l — APPROVED** [review]

File: tensorrt_llm/_torch/visual_gen/teacache.py [hunk 3]
```diff
@@ -115,25 +115,18 @@ def _extract_forward_args(self, module: torch.nn.Module, *args, **kwargs) -> Dic
...
-        # Call configured timestep embedding function
-        try:
-            return self.config.timestep_embed_fn(module, timestep_flat, guidance)
-        except Exception as e:
-            logger.error(f"Timestep embedder failed: {e}")
-            # Last resort: use timestep as-is
-            logger.warning("Using timestep fallback")
-            return timestep_flat.unsqueeze(-1) if timestep_flat.ndim == 1 else timestep_flat
+        # Call configured embedding function with full params
+        return self.config.timestep_embed_fn(module, params)
```
Comment (karljang) [inline tensorrt_llm/_torch/visual_gen/teacache.py]: Here, I’m passing all the parameters because FLUX requires additional tensors.  @o-stoner

**Review by QiJune — APPROVED** [review]
LGTM

**Review by chang-l — APPROVED** [review]
Can we also add `use_ret_steps` in the example, to align w/ WAN?

File: tensorrt_llm/_torch/visual_gen/teacache.py [hunk 4]
```diff
@@ -115,25 +115,16 @@ def _extract_forward_args(self, module: torch.nn.Module, *args, **kwargs) -> Dic
...
-
-        # Call configured timestep embedding function
-        try:
-            return self.config.timestep_embed_fn(module, timestep_flat, guidance)
-        except Exception as e:
-            logger.error(f"Timestep embedder failed: {e}")
-            # Last resort: use timestep as-is
-            logger.warning("Using timestep fallback")
-            return timestep_flat.unsqueeze(-1) if timestep_flat.ndim == 1 else timestep_flat
+        return self.config.timestep_embed_fn(module, **params)
```
[the remark this answers is not in the document]
Comment (karljang) [reply tensorrt_llm/_torch/visual_gen/teacache.py]: _We removed the flattening logic. The timestep_flat logic `(timestep.flatten() if timestep.ndim == 2 else timestep)` was only in the old centralized `_compute_timestep_embedding` in `teacache.py`. None of the actual embed callbacks (FLUX or WAN) ever flatten the timestep — they receive it directly from the scheduler which already provides it as a scalar/1D tensor.  So it was dead code in the old design._

And I'll add `use_ret_steps`~
